sycl/include/sycl/reduction.hpp

```diff
@@ -917,81 +918,81 @@ bool reduCGFuncForRangeFastReduce(handler &CGH, KernelType KernelFunc,
-    reductionLoop(Range, PerGroup, Reducer, NDId, KernelFunc);
+  auto Rest = [&](auto NWorkGroupsFinished) {
+    auto DoReducePartialSumsInLastWG =
+        Reduction::template getReadWriteLocalAcc<int>(1, CGH);
+
+    using Name = __sycl_reduction_kernel<reduction::main_krn::RangeFastReduce,
+                                         KernelName, decltype(NWorkGroupsFinished)>;
+    size_t PerGroup = Range.size() / NWorkGroups;
+    CGH.parallel_for<Name>(NDRange, [=](nd_item<1> NDId) {
+      // Call user's functions. Reducer.MValue gets initialized there.
+      typename Reduction::reducer_type Reducer;
+      reductionLoop(Range, PerGroup, Reducer, NDId, KernelFunc);
 
-    typename Reduction::binary_operation BOp;
-    auto Group = NDId.get_group();
+      typename Reduction::binary_operation BOp;
+      auto Group = NDId.get_group();
 
-    // If there are multiple values, reduce each separately
-    // reduce_over_group is only defined for each T, not for span<T, ...>
-    size_t LID = NDId.get_local_id(0);
-    for (int E = 0; E < NElements; ++E) {
-      auto &RedElem = Reducer.getElement(E);
-      RedElem = reduce_over_group(Group, RedElem, BOp);
-      if (LID == 0) {
-        if (NWorkGroups == 1) {
-          auto &OutElem = Reduction::getOutPointer(Out)[E];
-          // Can avoid using partial sum and write the final result immediately.
-          if (IsUpdateOfUserVar)
-            RedElem = BOp(RedElem, OutElem);
-          OutElem = RedElem;
-        } else {
-          PartialSums[NDId.get_group_linear_id() * NElements + E] =
-              Reducer.getElement(E);
+      // If there are multiple values, reduce each separately
+      // reduce_over_group is only defined for each T, not for span<T, ...>
+      size_t LID = NDId.get_local_id(0);
+      for (int E = 0; E < NElements; ++E) {
+        auto &RedElem = Reducer.getElement(E);
+        RedElem = reduce_over_group(Group, RedElem, BOp);
+        if (LID == 0) {
+          if (NWorkGroups == 1) {
+            auto &OutElem = Reduction::getOutPointer(Out)[E];
+            // Can avoid using partial sum and write the final result
+            // immediately.
+            if (IsUpdateOfUserVar)
+              RedElem = BOp(RedElem, OutElem);
+            OutElem = RedElem;
+          } else {
+            PartialSums[NDId.get_group_linear_id() * NElements + E] =
+                Reducer.getElement(E);
+          }
         }
       }
-    }
-
-    if (NWorkGroups == 1)
-      // We're done.
-      return;
 
-    // Signal this work-group has finished after all values are reduced
-    if (LID == 0) {
-      auto NFinished =
-          sycl::atomic_ref<int, memory_order::relaxed, memory_scope::device,
-                           access::address_space::global_space>(
-              NWorkGroupsFinished[0]);
-      DoReducePartialSumsInLastWG[0] = ++NFinished == NWorkGroups;
-    }
+      if (NWorkGroups == 1)
+        // We're done.
+        return;
```

Comment:
Any reason this happens so late? Is the accessor creation needed for dependency reasons?

Reply:
After offline discussion, I believe this question was caused by the confusion in the other comment. Disregard it.